File: internal/encode/languages.go
package encode

// Language is one entry in the curated common-language list: the normalized ISO
// 639-1 code stored in profiles and a human display name for the UI.
type Language struct {
	Code string `json:"code"`
	Name string `json:"name"`
}

// CommonLanguages is the curated list of selectable track languages, shared as
// the single source of truth for REST validation; the frontend mirrors it.
// Codes match the normalized output of normalizeLang. Adding a language is one
// entry here (plus its aliases in langAliases for tag normalization).
var CommonLanguages = []Language{
	{Code: "en", Name: "English"},
	{Code: "ja", Name: "Japanese"},
	{Code: "es", Name: "Spanish"},
	{Code: "pt", Name: "Portuguese"},
	{Code: "fr", Name: "French"},
	{Code: "de", Name: "German"},
	{Code: "it", Name: "Italian"},
	{Code: "ru", Name: "Russian"},
	{Code: "zh", Name: "Chinese"},
	{Code: "ko", Name: "Korean"},
	{Code: "ar", Name: "Arabic"},
}

// knownLanguageCodes indexes CommonLanguages by code for O(1) validation.
var knownLanguageCodes = func() map[string]bool {
	m := make(map[string]bool, len(CommonLanguages))
	for _, l := range CommonLanguages {
		m[l.Code] = true
	}
	return m
}()

// IsKnownLanguage reports whether a code is in the curated common-language list.
func IsKnownLanguage(code string) bool { return knownLanguageCodes[code] }

File: frontend/src/lib/languages.ts
// Curated common-language list for per-track audio/subtitle selection.
// Mirrors the Go source of truth `encode.CommonLanguages` (internal/encode/languages.go);
// codes match normalizeLang's output and are validated server-side (unknown → HTTP 400).
// Adding a language is one entry here plus the Go mirror.

export interface Language {
  code: string
  name: string
}

export const COMMON_LANGUAGES: readonly Language[] = [
  { code: 'en', name: 'English' },
  { code: 'ja', name: 'Japanese' },
  { code: 'es', name: 'Spanish' },
  { code: 'pt', name: 'Portuguese' },
  { code: 'fr', name: 'French' },
  { code: 'de', name: 'German' },
  { code: 'it', name: 'Italian' },
  { code: 'ru', name: 'Russian' },
  { code: 'zh', name: 'Chinese' },
  { code: 'ko', name: 'Korean' },
  { code: 'ar', name: 'Arabic' },
]

const NAME_BY_CODE = new Map(COMMON_LANGUAGES.map((l) => [l.code, l.name]))

export function languageName(code: string): string {
  return NAME_BY_CODE.get(code) ?? code
}
